scanner/src/main/kotlin/FileListing.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2023 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner
+
+import com.fasterxml.jackson.databind.annotation.JsonSerialize
+import com.fasterxml.jackson.databind.util.StdConverter
+
+import java.util.SortedSet
+
+import org.ossreviewtoolkit.scanner.FileListing.FileEntry
+import org.ossreviewtoolkit.utils.common.StringSortedSetConverter
+import org.ossreviewtoolkit.utils.common.getDuplicates
+
+/**
+ * The model to store file listings for provenances in a storage.
+ */
+internal data class FileListing(
+    /**
+     * A set of glob expressions that match directory paths which have been excluded from the file listing.
+     */
+    @JsonSerialize(converter = StringSortedSetConverter::class)
+    val ignoredDirectories: Set<String>,
+
+    @JsonSerialize(converter = FileEntrySortedSetConverter::class)
+    val files: Set<FileEntry>
+) {
+    data class FileEntry constructor(
+        val path: String,
+        val sha1: String
+    )
+
+    init {
+        val duplicates = files.getDuplicates { it.path }.keys
+
+        require(duplicates.isEmpty()) {
+            "The file listing contains duplicate paths which is not allowed: ${duplicates.joinToString()}."
+        }
+    }
```

Comment:
> Have you considered using a map with `path` as the key and hash(es) (or other file properties) as values to avoid the need to this check?

Yes I considered that. I didn't / don't find this check problematic - I thought the data structure as-is would be a tad more speaking.

Reply:
> Do you think a map would be better?

I'm unsure myself, esp. when it comes to how the serialization looks, but that's probably anyway unimportant here. I was more considering the case when more properties, like a Git-style SHA1, would be added. Maybe if would be nicer to have paths to files and files properties more clearly "separated" via key and values.

Reply:
Speaking of more properties to add, have you also considered to use (a set of) `Hash` classes instead of `String`s (and ensure that at least an entry with `HashAlgorithm.SHA1` is present)?

Reply:
> I'm unsure myself, esp. when it comes to how the serialization looks, but that's probably anyway unimportant here. I was more considering the case when more properties, like a Git-style SHA1, would be added. Maybe if would be nicer to have paths to files and files properties more clearly "separated" via key and values.

I believe a dedicated property is a "clear" separation already.
Adding further properties is out of scope of the this PR and the whole epic.
(This can be refactored quite easily if needed in the future)

Reply:
> Speaking of more properties to add, have you also considered to use (a set of) `Hash` classes instead of `String`s (and ensure that at least an entry with `HashAlgorithm.SHA1` is present)?

No, this is not needed for an MVP.

Reply:
> I believe a dedicated property is a "clear" separation already.

That depends on whether you regard the path to be a property of the file, like the SHA1. Or if you believe the path to be an identifier for the file, and the file (contents) have properties separate from that, like hashes, modification time, etc.

> Adding further properties is out of scope of the this PR and the whole epic.
> (This can be refactored quite easily if needed in the future)

It's certainly needed, and it also corresponds to what we discussed some time ago: Any implementation of this epic must be prepared for adding additional hashes (like the mentioned Git-style SHA1, for Software Heritage interoperability).

> No, this is not needed for an MVP IMO.

Obviously it's not technically needed. The question is whether it makes sense / is prescient to already implement it like that. Or phrased differently, why does it make sense *not* to use the `Hash` class here, since we already use it elsewhere for serialization.

Reply:
> It's certainly needed, and it also corresponds to what we discussed some time ago: Any implementation of this epic must be prepared for adding additional hashes (like the mentioned Git-style SHA1, for Software Heritage interoperability).

As said, it's easy as-is already to add further properties. So, yes it is prepared well enough IMO.

> Obviously it's not technically needed. The question is whether it makes sense / is prescient to already implement it like that. Or phrased differently, why does it make sense not to use the Hash class here, since we already use it elsewhere for serialization.

It does make sense to not do it because it is not needed in scope of my epic.

Reply:
> It does make sense to not do it because it is not needed in scope of my epic.

So is our policy now that everybody just strictly scratches their own itch, disregarding what other related use-cases others may have in the future, and leaving any additional work to them, even if it would not hurt to do it now? That's not the way at least I want to work on this project.